### PR TITLE
added current cluster id filter

### LINKED ISSUE #1431: Add ability to filter prometheus queries by current cluster ID
Now OpenCost queries do not have current cluster id filter as they expected to be performed against per-cluster Prometheus which has only current cluster metrics. In a setup with Victoria Metrics it requires to have agent per each cluster for metrics aggregations and this agent has no /metrics endpoint, it pushes metrics to monitoring cluster. It's inconvenient to have a separate Prometheus in each cluster just for OpenCost.
With current cluster selector it's possible to start using a shared monitoring cluster instead of a per-cluster Prometheus

### PR DESCRIPTION
**Pull request by AndrewChubatiuk**
Fixes #1431
## What does this PR change?
Introduces `CURRENT_CLUSTER_ID_FILTER_ENABLED` env variable which adds `<cluster_id_label>=<current_cluster_id>` filter to each prometheus query. It allows using shared prometheus for multiple Kubernetes clusters

## Does this PR relate to any other PRs?
* No

## How will this PR impact users?
* It just adds optional variable which by default doesn't change existing functionality

## How was this PR tested?
* It was tested in my EKS cluster with `CURRENT_CLUSTER_ID_FILTER_ENABLED` set to `false`, `true` and unset. Used UI for testing, so it means that tested `/allocation` endpoint only

## Does this PR require changes to documentation?
* `CURRENT_CLUSTER_ID_FILTER_ENABLED` variable can be added to documentation

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next Opencost release? If not, why not?
* Do not know which labels do you use

### CODE REVIEW

**Comment by AjayTripathy** [comment]
Hi @AndrewChubatiuk sorry for the late response here, we were at Kubecon. In general I'm supportive of these changes. Could you add a few more notes under "testing"? For example, was this tested successfully when no clusterID was specified as well as when one was? If so what endpoints were used to test?

**Comment by AndrewChubatiuk** [comment]
@AjayTripathy done

**Comment by Pokom** [comment]
@AndrewChubatiuk @AjayTripathy I'm curious what the state of the PR is. This is a feature I'd love to see merged in so I can remove the usage of `prom-label-proxy` for my setup. 

I'd be more then happy to help out with testing if need be.

**Comment by AjayTripathy** [comment]
Yeah-- I support this in general. Just a pretty large diff in a core module that requires careful review and testing. @Pokom if you grab the PR and test it out a third pair of eyes seems super useful!!

**Comment by Pokom** [comment]
Sounds good @AjayTripathy. I'm going to be off most of next week so I don't think I'll be able to get around to it until the new year.

I'd also be curious if we can leverage any internal code from `prom-label-proxy` so we can avoid having to rewrite all of our queries :grimacing: My thought would be to have some sort of flag(`ADD_CLUSTER_TO_PROMQL` or something similar, naming is hard) that we can check before sending queries. If it's true, we would append the cluster flag to the PromQL expression. All hand wavy, but I'd be happy to hack on it in my short time before the end of the year.

**Comment by AndrewChubatiuk** [comment]
hey @AjayTripathy
Are there any changes needed to merge it?

**Comment by AndrewChubatiuk** [comment]
@AjayTripathy rebased against develop

**Comment by AjayTripathy** [comment]
Thanks for the rebase @AndrewChubatiuk ... tagging for review and release in 103. @kaelanspatel would be the right person to review as its related to federation-- feel free to ping me @kaelanspatel and we can discuss.

File: pkg/costmodel/router.go
Comment (kaelanspatel) [inline pkg/costmodel/router.go]: Can you give some more context on the changes here? How does this relate to the filtering?

**Comment by kaelanspatel** [comment]
Hi @AndrewChubatiuk

Thanks for the PR, and for keeping it up to date. Tested the features myself and everything appears to be working. There is a little work required to make this Kubecost compatible (Assets runs queries on the closed-source backend), but that's really just a note for posterity and requires nothing on your end within this PR. As an Opencost feature I'm supportive of merging.

Just one issue with the changes to `router.go` above.

**Comment by AndrewChubatiuk** [comment]
> Hi @AndrewChubatiuk
> 
> Thanks for the PR, and for keeping it up to date. Tested the features myself and everything appears to be working. There is a little work required to make this Kubecost compatible (Assets runs queries on the closed-source backend), but that's really just a note for posterity and requires nothing on your end within this PR. As an Opencost feature I'm supportive of merging.
> 
> Just one issue with the changes to `router.go` above.

reverted router changes

**Review by kaelanspatel — DISMISSED** [review]
Thanks @AndrewChubatiuk, approved.

We will merge this once a follow up for those Opencost-Kubecost compatibility items is complete. Thanks again.

**Comment by AndrewChubatiuk** [comment]
I remembered a reason, why that changes in router.go were needed. We're using Victoria Metrics which consists of VMSingle and VMAgent in a basic setup. VMSingle has /api/v1/query endpoint and this is a part where a whole data aggregation happens and VMAgent is a per-cluster app which has /api/v1/status/config endpoint and it scrapes data in a cluster and pushes it to VMSingle.

Opencost expects both `/status/config` and `/query` endpoints to be present on the same host. In a `router.go` I've disabled `/status/config` to make opencost working, cause PROMETHEUS_SERVER_ENDPOINT was set to VMSingle address. I need to add another client which will point to PROMETHEUS_SERVER_ENDPOINT only if PROMETHEUS_SCRAPE_ENDPOINT is not set

**Comment by kaelanspatel** [comment]
Hey, thanks for the change.

If possible, could we move that to a different PR, just for bookkeeping's sake? It's not directly related to the filter work and requires some further verification, so I'd like to keep it separate.

**Comment by AndrewChubatiuk** [comment]
reverted the last change

**Review by kaelanspatel — APPROVED** [review]
Thanks very much.

Re-approved.

**Comment by mattray** [comment]
1.103 is coming out shortly, still good to merge @kaelanspatel after rebasing?

**Comment by kaelanspatel** [comment]
@mattray Yes, still ready to merge. Added that quick closed-source change to allow this to function with Kubecost at-large w.r.t. Assets, so we should be good.

**Comment by mattray** [comment]
@AndrewChubatiuk it looks like you're still pretty active on this PR. When you're done let's get the branch rebased with `develop` and we'll get it merged

**Comment by AndrewChubatiuk** [comment]
@mattray I've just rebased it against develop to keep it up to date. I'm not working on it

**Comment by mattray** [comment]
@kaelanspatel could you give this a final once-over and merge it?

**Comment by AndrewChubatiuk** [comment]
there were no changes since PR was approved. only rebased it several times

File: pkg/prom/diagnostics.go
Comment (skl) [inline pkg/prom/diagnostics.go]:
```suggestion
		Label:       "Expected cAdvisor labels available",
```

**Review by skl — COMMENTED** [review]
(nit typo spotted) looking forward to this! +1

**Comment by AndrewChubatiuk** [comment]
> (nit typo spotted) looking forward to this! +1

thank you, good catch 🙂

**Comment by kaelanspatel** [comment]
Hi all,

Sorry for the wait. Wanted to make sure this was all working properly, owing to the scale of the change; will be merging after a few more tests.

Tysm for keeping this up to date.

**Comment by kaelanspatel** [comment]
Thanks again all, will make sure this gets into the next release.

**Comment by AndrewChubatiuk** [comment]
@kaelanspatel thank you so much

**Pull request merged**
